api/rest/coordinator.go

```diff
@@ -0,0 +1,256 @@
+	workDir             string
+}
+
+// job is a function that will be called by the worker function. The output will be added to results channel
+type job func() BundleStatus
+
+// worker is a function that will run incoming jobs from jobs channel and put jobs output to results chan
+func worker(ctx context.Context, jobs <-chan job, results chan<- BundleStatus) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case j := <-jobs:
+			results <- j()
+		}
+	}
+}
+
+// NewParallelCoordinator constructs a new parallelcoordinator.
+func NewParallelCoordinator(client Client, interval time.Duration, workDir string) ParallelCoordinator {
+	return ParallelCoordinator{
+		client:              client,
+		statusCheckInterval: interval,
+		workDir:             workDir,
+	}
+}
+
+// CreateBundle starts the bundle creation process. Status updates be monitored
+// on the returned channel.
+func (c ParallelCoordinator) CreateBundle(ctx context.Context, id string, nodes []node) <-chan BundleStatus {
+
+	jobs := make(chan job)
+	statuses := make(chan BundleStatus)
+
+	for i := 0; i < numberOfWorkers; i++ {
+		go worker(ctx, jobs, statuses)
+	}
+
+	for _, n := range nodes {
+		logrus.WithField("IP", n.IP).Info("Sending creation request to node.")
+
+		// necessary to prevent the closure from giving the same node to all the calls
+		tmpNode := n
+		jobs <- func() BundleStatus {
+			select {
+			case <-ctx.Done():
+				return BundleStatus{
+					id:   id,
+					node: tmpNode,
+					err:  fmt.Errorf(contextDoneErrMsg),
+				}
+			default:
+			}
+
+			return c.createBundle(ctx, tmpNode, id, jobs)
+		}
+	}
+
+	return statuses
+}
+
+// CollectBundle waits until all the nodes' bundles have finished, downloads,
+// and merges them. The resulting bundle zip file path is returned.
+func (c ParallelCoordinator) CollectBundle(ctx context.Context, bundleID string, numBundles int, statuses <-chan BundleStatus) (string, error) {
+
+	// holds the paths to the downloaded local bundles before merging
+	var bundlePaths []string
+
+	finishedBundles := 0
+	for finishedBundles < numBundles {
+		select {
+		case <-ctx.Done():
+			return "", fmt.Errorf("context canceled before all node bundles finished")
+		case s := <-statuses:
+			if !s.done {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.id).Info("Got status update. Bundle not ready.")
+				continue
+			}
+			if s.err != nil {
+				logrus.WithError(s.err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Bundle errored")
+				// TODO (https://jira.mesosphere.com/browse/DCOS_OSS-5303): this should be noted in the generated bundle and not just printed in the journal
+				finishedBundles++
+				continue
+			}
+
+			bundlePath := filepath.Join(c.workDir, nodeBundleFilename(s.node))
+			err := c.client.GetFile(ctx, s.node.baseURL, s.id, bundlePath)
+			if err != nil {
+				logrus.WithError(err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Could not download file")
```

Comment:
I think we should `continue` on this step. If there is an error we may not be able to merge zip if it does not exist. We need a test case for this case
```suggestion
				logrus.WithError(err).WithField("IP", s.node.IP).WithField("ID", s.id).Warn("Could not download file")
				finishedBundles++
				continue
```